src/protocol/redpine_cc2500.c

```diff
@@ -0,0 +1,462 @@
+    packet[11] = 0;
+
+    u16 lcrc = crc(&packet[0], 12);
+    packet[12] = lcrc >> 8;
+    packet[13] = lcrc;
+
+}
+
+static u16 redpine_cb() {
+
+  switch(state) {
+    default:
+      set_start(47);
+      CC2500_SetPower(Model.tx_power);
+      CC2500_Strobe(CC2500_SFRX);
+      redpine_build_bind_packet();
+      CC2500_Strobe(CC2500_SIDLE);
+      CC2500_WriteData(packet, packet[0]+1);
+      state++;
+#ifndef EMULATOR
+      return 9000;
+#else
+      return 90;
+#endif
+    case REDPINE_BIND_DONE:
+      PROTOCOL_SetBindState(0);
+      initialize_data(0);
+      channr = 0;
+      state++;
+      break;
+    case REDPINE_DATA1:
+      if (fine != (s8)Model.proto_opts[PROTO_OPTS_FREQFINE]) {
+          fine = (s8)Model.proto_opts[PROTO_OPTS_FREQFINE];
+          CC2500_WriteReg(CC2500_0C_FSCTRL0, fine);
+      }
+      CC2500_SetTxRxMode(TX_EN);
+      set_start(channr);
+      CC2500_SetPower(Model.tx_power);
+      CC2500_Strobe(CC2500_SFRX);
+      redpine_data_frame();
+      CC2500_Strobe(CC2500_SIDLE);
+      CC2500_WriteData(packet, packet[0]+1);
+      channr = (channr + 1) % 47;
+      state = REDPINE_DATA1;
+#ifndef EMULATOR
+      return (Model.proto_opts[PROTO_OPTS_LOOPTIME]*100);
+#else
+      return Model.proto_opts[PROTO_OPTS_LOOPTIME];
+#endif
+  }
+  return 1;
+}
+
+// register, fast 250k, slow
+static const u8 init_data[][3] = {
+    {CC2500_00_IOCFG2,    0x06,  0x06},
+    {CC2500_02_IOCFG0,    0x06,  0x06},    
+    {CC2500_06_PKTLEN,    0x1E,  0x23},
+    {CC2500_07_PKTCTRL1,  0x04,  0x04},
+    {CC2500_08_PKTCTRL0,  0x05,  0x01},
+    {CC2500_0B_FSCTRL1,   0x0A,  0x08},
+    {CC2500_0C_FSCTRL0,   0x00,  0x00},
+    {CC2500_0D_FREQ2,     0x5D,  0x5c},
+    {CC2500_0E_FREQ1,     0x93,  0x80},
+    {CC2500_0F_FREQ0,     0xB1,  0x00},
+    {CC2500_10_MDMCFG4,   0x2D,  0x7B},
+    {CC2500_11_MDMCFG3,   0x3B,  0xF8},
+    {CC2500_12_MDMCFG2,   0x73,  0x03},
+    {CC2500_13_MDMCFG1,   0x22,  0x23},
+    {CC2500_14_MDMCFG0,   0xF8,  0x7a},
+    {CC2500_15_DEVIATN,   0x00,  0x53},
+    {CC2500_17_MCSM1,     0x0c,  0x0E},
+    {CC2500_18_MCSM0,     0x18,  0x18},
+    {CC2500_3E_PATABLE,   0xff,  0xff},
+
+};
+
+// register, value
+static const u8 init_data_shared[][2] = {
+    {CC2500_03_FIFOTHR,   0x07},
+    {CC2500_09_ADDR,      0x00},
+    {CC2500_19_FOCCFG,    0x1D},
+    {CC2500_1A_BSCFG,     0x1C},
+    {CC2500_1B_AGCCTRL2,  0xC7},
+    {CC2500_1C_AGCCTRL1,  0x00},
+    {CC2500_1D_AGCCTRL0,  0xB0},
+    {CC2500_21_FREND1,    0xB6},
+    {CC2500_22_FREND0,    0x10},
+    {CC2500_23_FSCAL3,    0xa9},
+    {CC2500_24_FSCAL2,    0x0A},
+    {CC2500_25_FSCAL1,    0x00},
+    {CC2500_26_FSCAL0,    0x11},
+    {CC2500_29_FSTEST,    0x59},
+    {CC2500_2C_TEST2,     0x88},
+    {CC2500_2D_TEST1,     0x31},
+    {CC2500_2E_TEST0,     0x0B},
+};
+
+static void redpine_init() {
+  CC2500_Reset();
+
+  u8 format = Model.proto_opts[PROTO_OPTS_FORMAT] + 1;
+
+  for (u32 i=0; i < ((sizeof init_data) / (sizeof init_data[0])); i++)
```

Comment:
use unsigned here. no need explicit on u32

Reply:
unsigned optimizes better?

Reply:
yes. only use u8/u16/u32 in the heap variables (global variables). for the stack variables, unsigned is better. or just int.